File: function/python_3_9/src/response/part_number_mapper.py
import math
from response import MapperResponse


DEFAULT_SIZE = 5242880


def get_part_error_response(number, total):
    """Get a MapperResponse for Errors"""
    return MapperResponse(hasError=True, object=None,
                          error_msg='Cannot specify part number: {}. Use part numbers 1 to {}.'.format(number, total))


def map_part_number(transformed_object, part_number):
    """
    Map the part number of an object
    :param transformed_object: transformed object
    :param part_number: part number request string
    :return: MapperResponse
    """
    object_length = len(transformed_object)
    total_part = math.ceil(object_length / DEFAULT_SIZE)
    try:
        part_number = int(part_number)
    except ValueError:
        return get_part_error_response(part_number, total_part)
    if part_number > total_part or part_number < 0:
        return get_part_error_response(part_number, total_part)

    start = (part_number - 1) * DEFAULT_SIZE
    end = min(start + DEFAULT_SIZE, object_length)
    return MapperResponse(hasError=False, object=transformed_object[start:end],
                          error_msg=None)

File: function/java11/src/main/java/com/example/s3objectlambda/transform/PartNumberMapper.java
package com.example.s3objectlambda.transform;

import com.example.s3objectlambda.exception.InvalidPartNumberException;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Arrays;

/**
 * Handles partNumber requests by applying the partNumber to the transformed object.
 */
public class PartNumberMapper {

    /*
    Minimum part size in a multipart upload is 5 MB.
    There is no size limit on the last part of the multipart upload.
     */
    private final Integer partSize = 5242880;
    private Logger logger;

    public PartNumberMapper() {
        this.logger = LoggerFactory.getLogger(PartNumberMapper.class);
    }

    /**
     * This method returns the requested part from the response object.
     * @param partNumber Part number , this should be >0 and <= Total number of parts in the response object.
     * @param responseObjectByteArray Response object as byte array, from which a particular part is requested.
     * @return Returns the byte array object of the requested part.
     * @throws InvalidPartNumberException
     */
    public byte[] mapPartNumber(String partNumber, byte[] responseObjectByteArray)
            throws InvalidPartNumberException {

        double objectLength;
        objectLength = responseObjectByteArray.length;
        double totalParts = Math.ceil(objectLength / this.partSize);
        int requestedPart;

        try {
            requestedPart = Integer.parseInt(partNumber);
        } catch (NumberFormatException nfe) {
            this.logger.error("Invalid partNumber" + nfe);
            throw new InvalidPartNumberException("Invalid partNumber: " + partNumber);
        }

        if (requestedPart > totalParts || requestedPart <= 0) {
            throw new InvalidPartNumberException(String.format("Cannot specify part number: %s. " +
                    "Use part number from 1 to %s.", requestedPart, totalParts));
        }

        int partStart = (requestedPart - 1) * this.partSize;
        int partEnd = (int) Math.min(partStart + this.partSize, objectLength);
        var objectPart = Arrays.copyOfRange(responseObjectByteArray, partStart, partEnd);
        return objectPart;

    }
}
